common/task/priorityTaskProcessor.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// PriorityTaskProcessorOptions configs PriorityTaskProcessor
+	PriorityTaskProcessorOptions struct {
+		WorkerCount int
+		RetryPolicy backoff.RetryPolicy
+	}
+
+	priorityTaskProcessorImpl struct {
+		scheduler PriorityScheduler
+
+		status       int32
+		workerWG     sync.WaitGroup
+		logger       log.Logger
+		metricsScope metrics.Scope
+		options      *PriorityTaskProcessorOptions
+	}
+)
+
+// NewPriorityTaskProcessor creates a new PriorityTaskProcessor
+func NewPriorityTaskProcessor(
+	scheduler PriorityScheduler,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	options *PriorityTaskProcessorOptions,
+) PriorityTaskProcessor {
+	return &priorityTaskProcessorImpl{
+		scheduler:    scheduler,
+		status:       common.DaemonStatusInitialized,
+		logger:       logger,
+		metricsScope: metricsClient.Scope(metrics.PriorityTaskProcessingScope),
+		options:      options,
+	}
+}
+
+func (p *priorityTaskProcessorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.workerWG.Add(p.options.WorkerCount)
+	for i := 0; i < p.options.WorkerCount; i++ {
+		go p.taskWorker()
+	}
+	p.scheduler.Start()
+	p.logger.Info("Priority task processor started.")
+}
+
+func (p *priorityTaskProcessorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	p.scheduler.Stop()
+	if success := common.AwaitWaitGroup(&p.workerWG, time.Minute); !success {
+		p.logger.Warn("Priority task processor timedout on shutdown.")
+	}
+	p.logger.Info("Priority task processor shutdown.")
+}
+
+func (p *priorityTaskProcessorImpl) Submit(task Task, priority int) error {
+	p.metricsScope.IncCounter(metrics.PriorityTaskSubmitRequest)
+	sw := p.metricsScope.StartTimer(metrics.PriorityTaskSubmitLatency)
+	defer sw.Stop()
+
+	return p.scheduler.Schedule(task, priority)
+}
+
+func (p *priorityTaskProcessorImpl) taskWorker() {
+	defer p.workerWG.Done()
+
+	for {
+		task := p.scheduler.Consume()
+		if task == nil {
+			return
+		}
+
+		p.executeTask(task.(Task))
+	}
+}
+
+func (p *priorityTaskProcessorImpl) executeTask(task Task) {
+	sw := p.metricsScope.StartTimer(metrics.PriorityTaskTaskProcessingLatency)
+	defer sw.Stop()
+
+	op := func() error {
+		if err := task.Execute(); err != nil {
+			return task.HandleErr(err)
+		}
+		return nil
+	}
+
+	isRetryable := func(err error) bool {
+		if p.isStopped() {
+			return false
+		}
+		return task.RetryErr(err)
+	}
+
+	for {
+		if p.isStopped() {
+			// neither ack or nack here
+			return
+		}
+
+		if err := backoff.Retry(op, p.options.RetryPolicy, isRetryable); err != nil {
+			if !task.RetryErr(err) {
+				// encounter an non-retryable error
+				task.Nack()
+				return
+			}
+			// retryable error, continue the loop
```

Comment:
is will do forever retry?

Reply:
I think here we rely on the implementation of the task interface to determine if retry is needed or not. The retry is only for transient errors like service busy error. For other errors like domain not active or shard ownership lost, the task should give up retry so that it can be Nacked. Inside `Nack()`, the task implementation can determine if the task should be discarded or re-dispatched.